[DOCS] Revamps transport docs (#1881)
elastic/elasticsearch-ruby: docs/transport.asciidoc

It does not handle calling the {es} or Enterprise Search APIs.

This library uses https://github.com/lostisland/faraday[Faraday] by default as the HTTP transport implementation. We test it it with Faraday versions 1.x and Faraday 2.x.

For optimal performance, use a HTTP library which supports persistent ("keep-alive") connections, such as https://github.com/toland/patron[patron] or https://github.com/typhoeus/typhoeus[Typhoeus]. Require the library (`require 'patron'`) in your code for Faraday 1.x or the adapter (`require 'faraday/patron'`) for Faraday 2.x, and it will be automatically used.

Currently these libraries are supported:

* https://github.com/toland/patron[Patron]
* https://github.com/typhoeus/typhoeus[Typhoeus]
* https://rubygems.org/gems/httpclient[HTTPClient]
* https://rubygems.org/gems/net-http-persistent[Net::HTTP::Persistent]

> **Note on https://github.com/typhoeus/typhoeus[Typhoeus]**: You need to use v1.4.0 or up since older versions are not compatible with Faraday 1.0.

Reviewer: NOTE is an admonition, so the statement will be a bit more prominent.

```suggestion
NOTE: Use https://github.com/typhoeus/typhoeus[Typhoeus] v1.4.0 or up since older versions are not compatible with Faraday 1.0.
```

> You can customize Faraday and implement your own HTTP transport. For detailed information, see the example configurations and more information [below](#transport-implementations).

Reviewer: ```suggestion
You can customize Faraday and implement your own HTTP transport. For detailed information, see the example configurations and more information <<transport-implementations,below>>.
```

# => #<Elastic::Transport::Transport::Response:0x007fc5d506ce38 @status=200, @body={ ... } >
----------------------------

Documentation is included as RDoc annotations in the source code and available online at http://rubydoc.info/gems/elastic-transport[RubyDoc].

[discrete]
[[transport-implementations]]
==== Transport implementations

By default, the client uses the https://rubygems.org/gems/faraday[Faraday] HTTP library as a transport implementation.

> It will auto-detect and use an _adapter_ for _Faraday_ based on gems loaded in your code, preferring HTTP clients with support for persistent connections. Faraday 2 changed the way adapters are used (https://github.com/lostisland/faraday/blob/main/UPGRADING.md#adapters-have-moved[read more here]). If you're using Faraday 1.x, you can require the HTTP library. To use the https://github.com/toland/patron[_Patron_] HTTP, for example, require it:

Reviewer: ```suggestion
The Client auto-detects and uses an _adapter_ for _Faraday_ based on gems loaded in your code, preferring HTTP clients with support for persistent connections. Faraday 2 changed the way adapters are used (https://github.com/lostisland/faraday/blob/main/UPGRADING.md#adapters-have-moved[read more here]). If you're using Faraday 1.x, you can require the HTTP library. To use the https://github.com/toland/patron[_Patron_] HTTP, for example, require it:
```

To use the https://github.com/toland/patron[Patron] HTTP, for example, require it:

[source,rb]
----------------------------
require 'patron'
----------------------------

> If you're using Faraday 2.x, you need to add the corresponding adapter gem to your Gemfile and require them after you require `faraday`:

Reviewer: ```suggestion
If you're using Faraday 2.x, you need to add the corresponding adapter gem to your Gemfile and require it after you require `faraday`:
```

> 
> ```ruby
> # Gemfile
> gem 'faraday-patron'
> 
> # Code
> require 'faraday'
> require 'faraday/patron'
> ```

Reviewer: ```suggestion

[source,rb]
----------------------------
# Gemfile
gem 'faraday-patron'

# Code
require 'faraday'
require 'faraday/patron'
----------------------------
```

Then, create a new client, and the Patron gem will be used as the "driver":

> ```ruby

Reviewer: ```suggestion
[source,rb]
----------------------------
```

# => [Faraday::Adapter::NetHttpPersistent]

> ```

Reviewer: ```suggestion
----------------------------
```

> If you see this error:

Reviewer: ```suggestion
If you see this error:

```

[source,rb]
----------------------------
Faraday::Error: :net_http_persistent is not registered on Faraday::Adapter
----------------------------

> When you're using Faraday 2, you need to require the adapter before instantiating the client:

Reviewer: ```suggestion
When you're using Faraday 2, you need to require the adapter before instantiating the client:

```

> ```ruby
> > client = Elasticsearch::Client.new(adapter: :net_http_persistent)
> Faraday::Error: :net_http_persistent is not registered on Faraday::Adapter
> > require 'faraday/net_http_persistent'
> => true
> > client = Elasticsearch::Client.new(adapter: :net_http_persistent)
> => #<Elasticsearch::Client:0x00007eff2e7728e0
> ```

Reviewer: ```suggestion
[source,rb]
----------------------------
> client = Elasticsearch::Client.new(adapter: :net_http_persistent)
Faraday::Error: :net_http_persistent is not registered on Faraday::Adapter
> require 'faraday/net_http_persistent'
=> true
> client = Elasticsearch::Client.new(adapter: :net_http_persistent)
=> #<Elasticsearch::Client:0x00007eff2e7728e0
----------------------------
```

* Logger and tracer can be any object conforming to Ruby logging interface, for example, an instance of https://ruby-doc.org/stdlib-1.9.3/libdoc/logger/rdoc/Logger.html[`Logger`], https://rubygems.org/gems/log4r[log4r], https://github.com/TwP/logging/[logging], and so on.
* The `Elastic::Transport::Transport::Serializer::Base` implementations handle converting data for {es} (for example, to JSON). You can implement your own serializer.
* `Elastic::Transport::Transport::Sniffer` allows to discover nodes in the cluster and use them as connections.
* `Elastic::Transport::Transport::Connections::Collection` is composed of `Elastic::Transport::Transport::Connections::Connection` instances and a selector instance.
* `Elastic::Transport::Transport::Connections::Connection` contains the connection attributes such as hostname and port, as well as the concrete persistent "session" connected to a specific node.
* The `Elastic::Transport::Transport::Connections::Selector::Base` implementations allow to choose connections from the pool, for example, in a round-robin or random fashion. You can implement your own selector strategy.

> * The `Elastic::Transport::Transport::Response` object wraps the Elasticsearch JSON response. It provides `body`, `status` and `headers` methods, but you can treat is as a hash and access the keys directly.

Reviewer: ```suggestion
* The `Elastic::Transport::Transport::Response` object wraps the Elasticsearch JSON response. It provides `body`, `status`, and `headers` methods but you can treat it as a hash and access the keys directly.
```
